File: crud-item-list/src/main/java/com/gurunars/crud_item_list/CollectionManager.java
package com.gurunars.crud_item_list;

import android.support.annotation.IdRes;
import android.view.View;

import com.esotericsoftware.kryo.Kryo;
import com.gurunars.item_list.Item;

import org.objenesis.strategy.StdInstantiatorStrategy;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import butterknife.ButterKnife;
import java8.util.function.Consumer;


class CollectionManager<ItemType extends Item> implements Serializable {

    private Kryo kryo = new Kryo();

    private List<ItemHolder<ItemType>> items = new ArrayList<>();
    private Set<ItemHolder<ItemType>> selectedItems = new HashSet<>();

    private final Consumer<List<SelectableItem<ItemType>>> stateChangeHandler;
    private final Consumer<List<ItemType>> collectionChangeHandler;
    private View contextualMenu;

    private final ActionEdit<ItemHolder<ItemType>> actionEdit = new ActionEdit<>();

    private final Map<View, Action<ItemHolder<ItemType>>> actions = new HashMap<>();

    CollectionManager(
            View contextualMenu,
            Consumer<List<SelectableItem<ItemType>>> stateChangeHandler,
            Consumer<List<ItemType>> collectionChangeHandler) {
        this.contextualMenu = contextualMenu;
        this.stateChangeHandler = stateChangeHandler;
        this.collectionChangeHandler = collectionChangeHandler;
        this.kryo.setInstantiatorStrategy(new Kryo.DefaultInstantiatorStrategy(new StdInstantiatorStrategy()));

        registerAction(R.id.selectAll, new ActionSelectAll<ItemHolder<ItemType>>());
        registerAction(R.id.delete, new ActionDelete<ItemHolder<ItemType>>());
        registerAction(R.id.edit, actionEdit);
        registerAction(R.id.moveUp, new ActionMoveUp<ItemHolder<ItemType>>());
        registerAction(R.id.moveDown, new ActionMoveDown<ItemHolder<ItemType>>());
    }

    private List<SelectableItem<ItemType>> getSelectableItems() {
        List<SelectableItem<ItemType>> rval = new ArrayList<>();
        for (ItemHolder<ItemType> item: items) {
            rval.add(new ConcreteSelectableItem<>(item.getRaw(), selectedItems.contains(item)));
        }
        return rval;
    }

    void registerAction(@IdRes int itemId, final Action<ItemHolder<ItemType>> action) {
        View itemView = ButterKnife.findById(contextualMenu, itemId);
        this.actions.put(itemView, action);
        itemView.setEnabled(action.canPerform(items, selectedItems));
        itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!action.canPerform(items, selectedItems)) {
                    return;
                }
                action.perform(items, selectedItems);
                changed();
                collectionChangeHandler.accept(ItemHolder.unwrap(items));
                v.setEnabled(action.canPerform(items, selectedItems));
            }
        });
    }

    boolean hasSelection() {
        return !selectedItems.isEmpty();
    }

    private void cleanSelection() {
        Set<ItemHolder<ItemType>> newSelection = new HashSet<>(selectedItems);
        for (ItemHolder<ItemType> cursor: selectedItems) {
            newSelection.remove(cursor);
            if (items.contains(cursor)) {
                // This is to replace the item's payload with a new one
                newSelection.add(items.get(items.indexOf(cursor)));
            }
        }
        selectedItems = newSelection;
    }

    private void changed() {
        cleanSelection();
        stateChangeHandler.accept(getSelectableItems());
        for (Map.Entry<View, Action<ItemHolder<ItemType>>> entry: actions.entrySet()) {
            entry.getKey().setEnabled(entry.getValue().canPerform(items, selectedItems));
        }
    }

    void itemClick(ItemType item) {
        if (selectedItems.size() == 0) {
            return;
        }

        ItemHolder<ItemType> holder = new ItemHolder<>(item);

        if (selectedItems.contains(holder)) {
            selectedItems.remove(holder);
        } else {
            selectedItems.add(holder);
        }
        changed();
    }

    void itemLongClick(ItemType item) {
        if (selectedItems.size() == 0) {
            selectedItems.add(new ItemHolder<>(item));
        }
        changed();
    }

    void setItemConsumer(final Consumer<ItemType> itemConsumer) {
        this.actionEdit.setItemConsumer(new Consumer<ItemHolder<ItemType>>() {
            @Override
            public void accept(ItemHolder<ItemType> holder) {
                itemConsumer.accept(holder.getRaw());
            }
        });
    }

    void unselectAll() {
        this.selectedItems.clear();
        changed();
    }

    void setItems(List<ItemType> items) {
        this.items = kryo.copy(new ArrayList<>(ItemHolder.wrap(items)));
        changed();
    }

    Serializable saveState() {
        return new HashSet<>(selectedItems);
    }

    void loadState(Serializable state) {
        selectedItems = (HashSet<ItemHolder<ItemType>>) state;
        changed();
    }

}

File: crud-item-list/src/main/kotlin/com/gurunars/crud_item_list/CollectionManager.kt
package com.gurunars.crud_item_list

import com.esotericsoftware.kryo.Kryo
import com.gurunars.item_list.Item
import org.objenesis.strategy.StdInstantiatorStrategy
import java.io.Serializable
import java.util.HashSet


internal class CollectionManager<ItemType : Item>(
        private val handleStateChange: (List<SelectableItem<ItemType>>) -> Unit,
        private val handleCollectionChange: (MutableList<ItemType>) -> Unit) : Serializable {

    private val kryo = Kryo()

    private var items = mutableListOf<ItemType>()
    private var selectedItems = mutableSetOf<ItemType>()
    private val selectedPositions : List<Int>
        get() = selectedItems.map {items.indexOf(it)}.sorted()

    private var itemConsumer: ((ItemType) -> Unit)? = null

    init {
        this.kryo.instantiatorStrategy = Kryo.DefaultInstantiatorStrategy(StdInstantiatorStrategy())
    }

    private val selectableItems: List<SelectableItem<ItemType>>
        get() {
            return items.map { ConcreteSelectableItem(it, selectedItems.contains(it)) }
        }

    fun hasSelection(): Boolean {
        return !selectedItems.isEmpty()
    }

    fun canEdit(): Boolean {
        return selectedItems.size == 1
    }

    fun isSolidChunk() : Boolean {
        val positions = selectedPositions
        if (positions.isEmpty()) {
            return false
        }
        var previousPosition = positions[0]

        return positions.subList(1, positions.size).all {
            val status = (it - previousPosition) == 1
            previousPosition = it
            return status
        }
    }

    fun canMoveUp(): Boolean {
        val positions = selectedPositions
        return isSolidChunk() && positions[0] > 0
    }

    fun canMoveDown(): Boolean {
        val positions = selectedPositions
        return isSolidChunk() && positions[positions.size - 1] < items.size - 1
    }

    fun canDelete(): Boolean {
        return !selectedItems.isEmpty()
    }

    fun canSelectAll(): Boolean {
        return selectedItems.size < items.size
    }

    private fun cleanSelection() {
        selectedItems = selectedItems.filter {items.contains(it)}.toMutableSet()
    }

    private fun changed() {
        cleanSelection()
        handleStateChange(selectableItems)
    }

    fun itemClick(item: ItemType) {
        if (selectedItems.size == 0) {
            return
        }
        if (selectedItems.contains(item)) {
            selectedItems.remove(item)
        } else {
            selectedItems.add(item)
        }
        changed()
    }

    fun itemLongClick(item: ItemType) {
        if (selectedItems.size == 0) {
            selectedItems.add(item)
        }
        changed()
    }

    fun setItemConsumer(itemConsumer: (ItemType) -> Unit) {
        this.itemConsumer = itemConsumer
    }

    fun unselectAll() {
        this.selectedItems.clear()
        changed()
    }

    fun setItems(items: List<ItemType>) {
        this.items = kryo.copy(items.toMutableList())
        changed()
    }

    private fun dataSetChanged() {
        changed()
        handleCollectionChange(items)
    }

    fun deleteSelected() {
        items = items.filter { !selectedItems.contains(it) }.toMutableList()
        dataSetChanged()
    }

    fun moveSelectionUp() {
        if (!canMoveUp()) {
            return
        }
        val positions = selectedPositions

        val positionToMoveDown = positions[0] - 1
        val itemToMoveDown = items[positionToMoveDown]

        items.add(positions[positions.size - 1] + 1, itemToMoveDown)
        items.removeAt(positionToMoveDown)

        dataSetChanged()
    }

    fun moveSelectionDown() {
        if (!canMoveDown()) {
            return
        }
        val positions = selectedPositions

        val positionToMoveUp = positions[positions.size - 1] + 1
        val itemToMoveUp = items[positionToMoveUp]

        items.removeAt(positionToMoveUp)
        items.add(positions[0], itemToMoveUp)

        dataSetChanged()
    }

    fun selectAll() {
        selectedItems.addAll(items)
        changed()
    }

    fun triggerConsumption() {
        if (itemConsumer == null) {
            return
        }

        val iterator = selectedItems.iterator()
        if (iterator.hasNext()) {
            itemConsumer?.invoke(kryo.copy(iterator.next()))
        }
    }

    fun saveState(): Serializable {
        return HashSet(selectedItems)
    }

    fun loadState(state: Serializable) {
        selectedItems = state as HashSet<ItemType>
        changed()
    }

}
